### ISSUE TITLE
Poor performance in VsTestV2 when using more test files (testAssemblyVer2)

### ISSUE THREAD

**Issue by jirikanda** [microsoft/azure-pipelines-task-lib#516]
### Environment
azure-pipelines-task-lib version: 2.*

### Issue Description
We are using more patterns to find assemblies with tests. For *every one pattern* the whole file system is traversed starting from the specified folder. This means for three patterns, three traversals are performed. For some projects, this is extremely slow.

#### More details
During build we do not set output folder, each project in the solution is compiled into SomeProject\bin\Release\... folder under the sources root. Traversal is slow for mainly traversing folders where tests should never be: 
* *.git* folder
* *packages* folder
* *SomeProject\node_modules* folder
 
### Expected behaviour
* Only one traversal should be performed.
* When it is sure the path cannot fullfill minimatch pattern, do not traverse subdirectories.

### Sources
The file system traversal is implemented in https://github.com/Microsoft/azure-pipelines-task-lib/blob/master/node/task.ts#L1585, critical is line 1585 (in the code block starting at line 1568). This is performed for each pattern in the for loop while it should be enough to traverse the file system once.

[11_Run Tests.txt](https://github.com/Microsoft/azure-pipelines-task-lib/files/3008933/11_Run.Tests.txt)

**Comment by hakenr** [comment]
Is there any chance to exclude specific folders (such as node_modules) from test assemblies discovery?

**Comment by BrandonBoone** [comment]
This seems to be stated in the OP's issue:

> When it is sure the path cannot fulfill minimatch pattern, do not traverse subdirectories.

But just in case it means something different, I would like to add the following:

If the minimatch pattern is prefixed with a relative path (*relative from the `searchFolder`*) or if the path is absolute (no minimatch patterns at all), short circuit and do not search for anything else in the `searchFolder` directories.

Example (pseudo config): 
`searchFolder: $(Build.SourcesDirectory)\src`

`testAssemblyVer2`
- `\Directory\Project\Tests\bin\*.tests.dll`
- `\Directory2\Project\Tests\bin\unittests.dll`

Even if there is a Directory3 through Directory100 and multiple subdirectories, etc. The only path that should be searched is `$(Build.SourcesDirectory)\src\Directory\Project\Tests\bin\` for `*.tests.dll`. `$(Build.SourcesDirectory)\src\Directory2\Project\Tests\bin\unittests.dll` should be found explicitly (no searching required).

I initially found this when using the [CopyFilesV2](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/CopyFilesV2/copyfiles.ts) task with absolute paths. I had to switch to using robocopy in a script block when I found out the entire source directory was being searched even though there were no actual patterns to match in my config.
